### PR TITLE
Fix: the trash post action doesn't take into account user capabilities.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
Right now we show the trash post action all the time even if the user does not have the correct capabilities.
We essentially regressed on https://github.com/WordPress/gutenberg/issues/23144.
This PR adds equivalent checks to the ones added at https://github.com/WordPress/gutenberg/pull/23174.

It is a bit hacky because we are calling a selector inside isEligible and to make sure things rerender we make the memorization depend on the result of getCachedResolvers.

It is also not ideal in the sense we make a REST API request per post to make sure the user has the permissions. In a list with 50 posts, it is not a very good solution.

I considered not using canUser, and instead retrieving the capabilities of the post and the user: delete_others_posts, delete_post, delete_posts, delete_private_posts, delete_published_posts,delete_published_posts. Implement the verification on the client e.g.: check if the post is published or private, and verify if the user capabilities are there. These would only depend on postType and user so we would not require a request per post. However, it is incompatible with some ways developers change capabilities (e.g.: the sample below would not work with this approach).

So I opted for the more reliable approach that we were already using.

This should be part of 6.6 as showing the UI to trash posts when the user has no capability for that is a regression.

cc: @youknowriad, @ntsekouras as people inside the actions work.

<b>Reviewers note:</b>
Hide whitespace: https://github.com/WordPress/gutenberg/pull/62589/files?w=1

The only changes are the inclusion of useRegistry and useSelect and their use inside isEligible, excluding these changes the action was kept exactly as before.

## Testing Instructions
I pasted the following code snipped on load.php:
```
function remove_delete_post_capability( $caps, $cap, $user_id, $args ) {
    if ( 'delete_post' == $cap ) {
        // Get the current user object
        $current_user = wp_get_current_user();
        // Check if the current user is the user we want to remove the capability from
        if ( $current_user->ID == $user_id ) {
            // Remove the 'delete_post' capability
            // It's done by returning a capability that the user does not have, for example 'do_not_allow'
            $caps[] = 'do_not_allow';
        }
    }

    return $caps;
}

add_filter( 'map_meta_cap', 'remove_delete_post_capability', 10, 4 );
```

I verified the option to trash pages was not available on Dataviews pages list and site editor.
I opened a post on the post editor and verified that the action to trash the post was unavailable.

### CODE REVIEW

File: packages/editor/src/components/post-actions/actions.js [hunk 1]
```diff
@@ -147,165 +147,207 @@ const deletePostAction = {
...
-								promiseResult.every(
-									( { status } ) => status === 'fulfilled'
-								)
-							) {
-								let successMessage;
-								if ( promiseResult.length === 1 ) {
-									successMessage = sprintf(
-										/* translators: The item's title. */
-										__( '"%s" moved to trash.' ),
+function useTrashPostAction( postType ) {
```
Comment (youknowriad) [inline packages/editor/src/components/post-actions/actions.js:310]: I don't like where this is going. I understand why this was done but I like to think actions should be global objects that don't require a hook to define them. This is going in the opposite direction of all my extensibility PRs.

Comment (jorgefilipecosta) [reply packages/editor/src/components/post-actions/actions.js:310]: Hi @youknowriad, this seems important to address on WP 6.6, we will be showing the trash button on the post editor on posts a user has no permission to trash, which is a considerable regression. As of right now there is no reliable way to fix the issue without using a hook, as soon as our APIs allow it and we can fix the issue without using a hook I'm happy to iterate and remove the hook here.

File: packages/editor/src/components/post-actions/actions.js [hunk 2]
```diff
@@ -1020,6 +1062,7 @@ export function usePostActions( { postType, onActionPerformed, context } ) {
 	);
 
 	const duplicatePostAction = useDuplicatePostAction( postType );
+	const trashPostAction = useTrashPostAction( postType );
```
Comment (ellatrix) [inline packages/editor/src/components/post-actions/actions.js]: Why don't we check dynamically add the action here instead of trying to check for the capabilities in `isEligible`?

File: packages/editor/src/components/post-actions/actions.js (hunk 1 above)
Comment (youknowriad) [reply packages/editor/src/components/post-actions/actions.js:310]: Indeed 6.6 is close, so I'm totally fine with this fix for 6.6 but we should definitely look at better APIs for actions after that.

File: packages/editor/src/components/post-actions/actions.js (hunk 2 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/post-actions/actions.js]: Hi @ellatrix, because a user may be able to trash a post and not another one. The permissions depend on the specific post. For example, a user maybe have permissions to delete only drafs, or permission to delete drafs authored by that user or permissions to delete private posts but not public ones, etc... We need to check for the specific post in isEligible.

File: packages/editor/src/components/post-actions/actions.js [hunk 3]
```diff
@@ -1113,6 +1143,9 @@ export function usePostActions( { postType, onActionPerformed, context } ) {
 		isPattern,
 		postTypeObject?.viewable,
 		duplicatePostAction,
+		permanentlyDeletePostAction,
+		trashPostActionForPostType,
+		restorePostAction,
```
Comment (ellatrix) [inline packages/editor/src/components/post-actions/actions.js]: @jorgefilipecosta Any reason for these extra unneeded dependencies?

File: packages/editor/src/components/post-actions/actions.js [hunk 4]
```diff
@@ -307,6 +307,35 @@ const trashPostAction = {
...
+				return (
+					trashPostAction.isEligible( item ) &&
+					registry
+						.select( coreStore )
+						.canUser( 'delete', resource, item.id )
+				);
+			},
+		} ),
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ registry, resource, canUserResolvers ]
```
Comment (ellatrix) [inline packages/editor/src/components/post-actions/actions.js]: Why do we need `canUserResolvers` here?

**Comment by ellatrix** [comment]
I minimised the diff, so it's easier to review and less likely to have cherry-pick conflicts.

File: packages/editor/src/components/post-actions/actions.js [hunk 5]
```diff
@@ -1048,7 +1078,7 @@ export function usePostActions( { postType, onActionPerformed, context } ) {
 			isTemplateOrTemplatePart ? resetTemplateAction : restorePostAction,
 			isTemplateOrTemplatePart || isPattern
 				? deletePostAction
```
Comment (ellatrix) [inline packages/editor/src/components/post-actions/actions.js:1082]: Don't we need the same for `deletePostAction`?

File: packages/editor/src/components/post-actions/actions.js [hunk 6]
```diff
@@ -161,186 +321,13 @@ function useTrashPostAction( postType ) {
 	);
 	return useMemo(
 		() => ( {
-			id: 'move-to-trash',
-			label: __( 'Move to Trash' ),
-			isPrimary: true,
-			icon: trash,
+			...trashPostAction,
 			isEligible( item ) {
```
Comment (youknowriad) [inline packages/editor/src/components/post-actions/actions.js:325]: The problem is that the `isEligible` call here won't be re-called when the value of `canUser( 'delete', resource, item.id )` changes no?

File: packages/editor/src/components/post-actions/actions.js (hunk 3 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/post-actions/actions.js]: It was probably an issue during a rebase I think we can remove them.

File: packages/editor/src/components/post-actions/actions.js (hunk 6 above)
Comment (ellatrix) [reply packages/editor/src/components/post-actions/actions.js:325]: Yeah, maybe we need something like `useEligible` somewhere. Or maybe this object should define which capabilities to check.

Comment (youknowriad) [reply packages/editor/src/components/post-actions/actions.js:325]: Related discussion https://github.com/WordPress/gutenberg/pull/62505#discussion_r1646480478

File: packages/editor/src/components/post-actions/actions.js (hunk 4 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/post-actions/actions.js]: When we do registry.select( coreStore ).canUser( 'delete', resource, item.id ), on the first call it will return false, then the selector starts returning true but nothing is causing a rerender, so the trash action would not be available until a rerender happens e.g: the user closes and reopen the sidebar. We make the memo depend on canUserResolvers as a way to cause a rerender when the resolver finishes. Of course, this is a temporary hacky fix, our actions API needs to offer a long-term solution as we are discussing in https://github.com/WordPress/gutenberg/pull/62647#discussion_r1644336613.

File: packages/editor/src/components/post-actions/actions.js (hunk 5 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/post-actions/actions.js:1082]: Yes, we need the same for some other actions, my plan was to define the fix here and after we agree on a fix I will replicate it to the other cases where it makes sense.

File: packages/editor/src/components/post-actions/actions.js (hunk 6 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/post-actions/actions.js:325]: It will because we make the memo depend on canUserResolvers (details in https://github.com/WordPress/gutenberg/pull/62589#discussion_r1647524094) so when the resolver finishes a rerender happens and isEligible will be recalled.

Comment (ellatrix) [reply packages/editor/src/components/post-actions/actions.js:325]: 🤨 seems fine for now but not ideal

Comment (youknowriad) [reply packages/editor/src/components/post-actions/actions.js:325]: It's very hard to follow. I'm not sure I understand what `canUserResolvers` is?

File: packages/editor/src/components/post-actions/actions.js (hunk 4 above)
Comment (ellatrix) [reply packages/editor/src/components/post-actions/actions.js]: Could you add an inline comment?

File: packages/editor/src/components/post-actions/actions.js (hunk 6 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/post-actions/actions.js:325]: canUserResolvers is the result getCachedResolvers().canUser. So basically when there is a change in the caching of canUser resolvers we will rerender making isEligible recalled. I will add some inline comments, explaining what we discussed.

File: packages/editor/src/components/post-actions/actions.js (hunk 4 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/post-actions/actions.js]: Hi @ellatrix I added an inline comment as suggested 👍

File: packages/editor/src/components/post-actions/actions.js (hunk 5 above)
Comment (ellatrix) [reply packages/editor/src/components/post-actions/actions.js:1082]: Is permanently delete a separate capability? I'm still seeing this:

image

Comment (ellatrix) [reply packages/editor/src/components/post-actions/actions.js:1082]: I tested with the testing instructions, and it seems to be the same capability: I can also not delete it permanently in trunk.

**Review by ellatrix — APPROVED** [review]
Not ideal, but we're not exposing any public APIs.

File: packages/editor/src/components/post-actions/actions.js (hunk 6 above)
Comment (ntsekouras) [reply packages/editor/src/components/post-actions/actions.js:325]: Also why keeping the action as object just to be used inside the `hook` action? Anyway since it's for 6.6, but it becomes clear that an action needs a way to have access to the store in more places than the callback and we should work towards that direction.

Comment (ellatrix) [reply packages/editor/src/components/post-actions/actions.js:325]: @ntsekouras I did that to minimise the diff, it was huge and unclear what changed + less likely to have cherry pick conflicts, and if it does, easier to resolve manually (I don't know what has changed in these files recently)

Comment (ellatrix) [reply packages/editor/src/components/post-actions/actions.js:325]: And it's useful for reusability https://github.com/WordPress/gutenberg/pull/62754#discussion_r1649305615

**Comment by ellatrix** [comment]
Handled in #62823.
